File: haskell/Simple/Wind.hs
module Simple.Wind ( wind ) where

import Variables
import Models
import Dataset

wind :: Dataset -> Model2D
wind dataset x@(Position lat _ alt) t =
    let (wind_u, wind_v) = interpolate dataset x t
        r = 6371009 + alt
        dlat = 180 / pi * wind_v / r
        dlon = 180 / pi * wind_u / (r * cos (lat * pi / 180))
    in
        Delta2D dlat dlon

File: tawhiri/models/wind.go
package models

import (
    "math"

    "../../tawhiri"
    "../dataset"
)

type Wind struct {
    DS dataset.Dataset
}

const EarthRadius = 6371009.0

func (w Wind) Eval(s tawhiri.State) (d tawhiri.Delta) {
    wind_u, wind_v := w.DS.Interpolate(s.Now, s.Lat, s.Lon, s.Alt)
    r := EarthRadius + s.Alt
    d.DLat = 180 / math.Pi * wind_v / r
    d.DLon = 180 / math.Pi * wind_u / (r * math.Cos(s.Lat*math.Pi/180))
    return
}
